### PR TITLE
X3d

### PR DESCRIPTION
**Pull request by wangyang59**
Initial draft of integrating xworld environments into flare. 
Please have a look at the general structure first. I will add tests later on.

### CODE REVIEW

File: flare/framework/env.py [hunk 1]
```diff
@@ -1,25 +1,163 @@
...
 """
 A file for wrapping different environments to have a unified API interface
 used by Agent's control flow.
 """
 
 
-class GymEnv(object):
+class Env(object):
+    """
+    An abstract class for environment. A new environment should inherit this
```
Comment (yu239) [inline flare/framework/env.py]: inherit -> inherit from

File: flare/framework/env.py [hunk 2]
```diff
@@ -1,25 +1,163 @@
...
+
+        return screen, sentence
+
+    def reset(self):
+        ## should return a list of observations
+        self.env.reset_game()
+        init_ob = self.env.get_state()
+        screen, sentence = self.preprocess_observation(init_ob)
+
+        self.buf = [np.zeros(screen.shape).astype(screen.dtype) \
```
Comment (yu239) [inline flare/framework/env.py]: xworld by default has an option of 'context'. So probably you don't need to maintain here.

File: flare/framework/env.py [hunk 3]
```diff
@@ -1,25 +1,163 @@
...
+    def __init__(self, game_name, contexts=1, options=None, dict=None):
+        """
+        options: see test_xworld3d.py in XWorld for an example of options
+        dict: dictionary for the langauge part. key=id, value=word
+        """
+        self.env = Simulator.create(game_name, options)
+        assert contexts >= 1, "contexts must be a positive number!"
+        self.contexts = contexts
+        self.buf = None
+        self.dict_id2w = dict
```
Comment (yu239) [inline flare/framework/env.py]: Only a list is enough for converting id to word

File: flare/framework/env.py [hunk 4]
```diff
@@ -1,25 +1,163 @@
...
+
+        total_reward = self.env.take_actions({
+            "action": a,
+            "pred_sentence": sentence
+        }, actrep, False)
+
+        next_game_over = self.env.game_over() != "alive"
+        next_ob = self.env.get_state()
+
+        screen, sentence = self.preprocess_observation(next_ob)
```
Comment (yu239) [inline flare/framework/env.py]: If the option 'context' is set, you don't have to maintain a buffer.

File: flare/framework/env.py [hunk 5]
```diff
@@ -49,6 +187,7 @@ def step(self, actions, actrep=1):
         total_reward = 0
         next_game_over = False
         for i in range(actrep):
+            self.steps += 1
```
Comment (yu239) [inline flare/framework/env.py]: This should be moved outside of the for loop.

File: flare/framework/env.py [hunk 6]
```diff
@@ -1,25 +1,163 @@
...
+
+
+class GymEnv(Env):
     """
     A convenient wrapper for OpenAI Gym env. Used to unify the
     environment interfaces.
     """
 
-    def __init__(self, game_name, contexts=1):
+    def __init__(self, game_name, contexts=1, options=None):
```
Comment (yu239) [inline flare/framework/env.py]: delete 'options=None'

File: flare/framework/env.py [hunk 7]
```diff
@@ -1,25 +1,163 @@
...
+    An abstract class for environment. A new environment should inherit this
+    class
+    """
+
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def reset(self):
+        """
+        reset the environment and return the initial observation
```
Comment (yu239) [inline flare/framework/env.py]: return a list of initial observations

File: flare/framework/env.py [hunk 8]
```diff
@@ -1,25 +1,163 @@
 import gym
 import numpy as np
 from scipy import misc
+from abc import ABCMeta, abstractmethod
+from py_simulator import Simulator
```
Comment (yu239) [inline flare/framework/env.py]: This might cause errors if the user hasn't installed py_simulator yet. Probably we should put the import inside the class?

**Comment by wangyang59** [comment]
@yu239 I have modified the codes according to your initial reviews. I have moved gym_env and xworld env into a separate folder of `env_zoo` and added a test for xworld_env

File: flare/framework/env.py [hunk 9]
```diff
@@ -45,10 +41,9 @@ def observation_dims(self):
     @abstractmethod
     def action_dims(self):
         """
-        Return a list of integers as action dimensions, each integer for an action.
-        For each integer,
-        if the corresponding action is discrete, then it means
-        the total number of actions;
+        Return a tuple of integers as action dimensions, each integer for an
```
Comment (yu239) [inline flare/framework/env.py]: tuple -> list. Use list for elements of different kinds

**Comment by wangyang59** [comment]
@yu239 I have rebased and changed the code accordingly.

**Pull request merged**